src/coord/tests/builtins.rs

```diff
@@ -0,0 +1,46 @@
+use coord::catalog::builtin::{
+    Builtin, BuiltinFunc, BuiltinLog, BuiltinTable, BuiltinView, BUILTINS,
+};
+use expr::GlobalId;
+use std::collections::BTreeSet;
+
+#[test]
+/// Ensures that all our builtins' fields are assigned globally-unique IDs.
+fn builtins_have_unique_ids() {
+    let mut encountered = BTreeSet::<GlobalId>::new();
+    let mut encounter =
+        move |kind_name: &str, field_name: &str, identifier: &str, id: &GlobalId| {
+            assert!(
+                encountered.insert(id.to_owned()),
+                "{} for {} {:?} is already used as a global ID: {:?}",
+                field_name,
+                kind_name,
+                identifier,
+                id,
+            );
+        };
+
+    use Builtin::*;
+    for (id, b) in BUILTINS.iter() {
+        match b {
+            Type(_) => {
+                let name = format!("with ID {:?}", id);
+                encounter("type", "id", &name, id);
+            }
+            Log(BuiltinLog { name, index_id, .. }) => {
+                encounter("type", "id", name, id);
+                encounter("type", "index_id", name, index_id);
+            }
+            Table(BuiltinTable { index_id, name, .. }) => {
+                encounter("builtin table", "id", name, id);
+                encounter("builtin table", "index_id", name, index_id);
+            }
+            View(BuiltinView { name, .. }) => {
+                encounter("view", "id", name, id);
+            }
+            Func(BuiltinFunc { name, .. }) => {
+                encounter("function", "id", name, id);
+            }
+        }
+    }
+}
```

Comment:
Thanks very much for doing this! I have a suggestion, though! Can we perform the check inline when we construct the `BUILTINS` variable? As in, somewhere here: https://github.com/MaterializeInc/materialize/blob/476f8b593e8640870101528cac7766880b3deb7d/src/coord/src/catalog/builtin.rs#L1336

The idea being that this test will catch the duplicate ID bug if someone is running `cargo test`, but if they're testing their change via `cargo run`, I think they'd still be doomed to bad error messages?

Reply:
Yep, that makes sense - I guess it will cause tests to fail too, incidentally (not sure how lazy_static will do its thing), so if things fall the way I imagine they will, I'll leave a test (but not this test) in there to ensure it triggers in CI, too.

Reply:
(Now that I think about it, there's probably no realistic branch of the universe on which our CI doesn't set up the builtin lazy_static initialization. I'll drop the test) (:

Reply:
Yup exactly. I can imagine other situations where you'd want to write a test like

```
#[test]
fn test_touch_static() {
  black_box(&*BUILTINS)
}
```

but here I agree that it's quite safe to rely on the many tests that actually start Materialize to run the BUILTINS initializer.